astropy/wcs/src/util.c

```diff
@@ -20,18 +20,28 @@ void set_invalid_to_nan(
   const int* s = stat;
   const int* s_end = stat + ncoord;
   double n;
+  int bit;
 
   #ifndef NAN
     #define INF (DBL_MAX+DBL_MAX)
     #define NAN (INF-INF)
   #endif
 
+  // Note that stat is a bit mask, so we need to mask only some of
+  // the coordinates depending on the bit mask values.
+
   n = NAN;
 
   for ( ; s != s_end; ++s) {
     if (*s) {
+      bit = 1;
       for (i = 0; i < nelem; ++i) {
-        *d++ = n;
+        if ((*s & bit) == bit) {
+          *d++ = n;
+        } else {
+          d++;
+        }
+        bit *= 2;
```

Comment:
I would say a left-shift would be more comprehensible here.

Reply:
Also do we need to check for overflow here? I'm a bit hazy on the details but signed integer overflow could lead to undefined behavior, right?

Reply:
I've emailed Mark Calabretta to ask whether there is a maximum supported number of dimensions in WCSLIB, to see if we can assume we are safe against overflow.

Reply:
I changed to left shift

Reply:
At least for NumPy the maximum number of dimensions is [32](https://github.com/numpy/numpy/blob/v1.17.4/numpy/core/include/numpy/ndarraytypes.h#L40) which could overflow a signed 32bit-integer (even if extremely unlikely):

```python
>>> import numpy as np
>>> np.array([[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[1]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]).ndim
32
```

Reply:
This is not a risk, naxis is limited to <=15:

```
In [2]: wcs = WCS(naxis=50)                                                     
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-9c7a363c3ea0> in <module>
----> 1 wcs = WCS(naxis=50)

~/Code/astropy/astropy/wcs/wcs.py in __init__(self, header, fobj, key, minerr, relax, naxis, keysel, colsel, fix, translate_units, _do_set)
    366                 naxis = 2
    367             wcsprm = _wcs.Wcsprm(header=None, key=key,
--> 368                                  relax=relax, naxis=naxis)
    369             self.naxis = wcsprm.naxis
    370             # Set some reasonable defaults.

ValueError: naxis must be in range 1-15
```

I added a comment about this.

Reply:
Thank you